### PR TITLE
Erase database data [11090]

### PR DESCRIPTION
**Pull request by JLBuenoLopez-eProsima**
This PR implements and tests `Database::erase` method.

### CODE REVIEW

File: src/cpp/database/database.cpp [hunk 1]
```diff
@@ -1352,6 +1352,71 @@ std::vector<std::pair<EntityId, EntityId>> Database::get_entities_by_name(
...
+        for (auto locator : reader.second.get()->locators)
+        {
+            locators_[locator.first].get()->data_readers.erase(reader.first);
+        }
+        // Clear entity
+        reader.second.get()->DDSEndpoint::clear();
+        reader.second.get()->clear();
+    }
+    // Clear datareaders map
+    datareaders_[domain_id].clear();
```
Comment (IkerLuengo) [inline src/cpp/database/database.cpp:1378]: Should remove the entry with key `domain_id` in the `datareaders_` map too, not just `clear` it. Same in the rest of entities

File: src/cpp/database/database.cpp [hunk 2]
```diff
@@ -1352,6 +1352,71 @@ std::vector<std::pair<EntityId, EntityId>> Database::get_entities_by_name(
...
+        topic.second.get()->clear();
+    }
+    // Clear topics map
+    topics_[domain_id].clear();
+
+    // Remove participants and unlink related process
+    for (auto participant : participants_[domain_id])
+    {
+        // Unlink related process
+        processes_[participant.second.get()->process.get()->id].get()->participants.erase(participant.first);
```
Comment (IkerLuengo) [inline src/cpp/database/database.cpp:1408]: What if the participant is not linked to any process?

**Review by IkerLuengo — CHANGES_REQUESTED** [review]

File: test/unittest/Database/DatabaseEraseTests.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    EXPECT_EQ(final_dump[USER_CONTAINER_TAG], erased_dump[USER_CONTAINER_TAG]);
+    EXPECT_EQ(final_dump[PROCESS_CONTAINER_TAG], erased_dump[PROCESS_CONTAINER_TAG]);
+    EXPECT_EQ(final_dump[DOMAIN_CONTAINER_TAG], erased_dump[DOMAIN_CONTAINER_TAG]);
+    EXPECT_EQ(final_dump[TOPIC_CONTAINER_TAG], erased_dump[TOPIC_CONTAINER_TAG]);
+    EXPECT_EQ(final_dump[PARTICIPANT_CONTAINER_TAG], erased_dump[PARTICIPANT_CONTAINER_TAG]);
+    EXPECT_EQ(final_dump[LOCATOR_CONTAINER_TAG], erased_dump[LOCATOR_CONTAINER_TAG]);
+    EXPECT_EQ(final_dump[DATAWRITER_CONTAINER_TAG], erased_dump[DATAWRITER_CONTAINER_TAG]);
+    EXPECT_EQ(final_dump[DATAREADER_CONTAINER_TAG], erased_dump[DATAREADER_CONTAINER_TAG]);
+
+    EXPECT_EQ(final_dump, erased_dump);
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/unittest/Database/DatabaseEraseTests.cpp:97]: It is legacy from the `load_and_dump` tests. I will remove also these redundant checks in those tests.

File: src/cpp/database/database.cpp (hunk 2 above)
Comment (JLBuenoLopez-eProsima) [reply src/cpp/database/database.cpp:1408]: Good catch. I think I should add a test checking this case :facepalm:

File: src/cpp/database/database.cpp [hunk 4]
```diff
@@ -1352,6 +1352,71 @@ std::vector<std::pair<EntityId, EntityId>> Database::get_entities_by_name(
     return entities;
 }
 
+void Database::erase(
+        EntityId& domain_id)
+{
+    // Check that the given domain_id corresponds to a known monitor.
+    // Upper layer ensures that the monitor has been stopped.
+    // Upper layer also ensures that the monitor_id is valid and corresponds to a known domain.
+    assert(EntityKind::DOMAIN == get_entity_kind(domain_id));
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/database/database.cpp]: Done in 14e23ec6c8db41f4b2f07fb7021c3931c3dbc05a

File: test/unittest/Database/DatabaseEraseTests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/unittest/Database/DatabaseEraseTests.cpp]: Done in 64290fb3affdb51394fdabf19a82ff43449e052f

File: test/unittest/Database/DatabaseLoadTests.cpp [hunk 5]
```diff
@@ -151,6 +152,40 @@ TEST(database_load_tests, load_and_dump_old_complex_database)
...
+    ASSERT_EQ(dump[USER_CONTAINER_TAG], loadedDump[USER_CONTAINER_TAG]);
+    ASSERT_EQ(dump[PROCESS_CONTAINER_TAG], loadedDump[PROCESS_CONTAINER_TAG]);
+    ASSERT_EQ(dump[DOMAIN_CONTAINER_TAG], loadedDump[DOMAIN_CONTAINER_TAG]);
+    ASSERT_EQ(dump[TOPIC_CONTAINER_TAG], loadedDump[TOPIC_CONTAINER_TAG]);
+    ASSERT_EQ(dump[PARTICIPANT_CONTAINER_TAG], loadedDump[PARTICIPANT_CONTAINER_TAG]);
+    ASSERT_EQ(dump[LOCATOR_CONTAINER_TAG], loadedDump[LOCATOR_CONTAINER_TAG]);
+    ASSERT_EQ(dump[DATAWRITER_CONTAINER_TAG], loadedDump[DATAWRITER_CONTAINER_TAG]);
+    ASSERT_EQ(dump[DATAREADER_CONTAINER_TAG], loadedDump[DATAREADER_CONTAINER_TAG]);
+
+    ASSERT_EQ(dump, loadedDump);
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/unittest/Database/DatabaseLoadTests.cpp]: Done in 64290fb3affdb51394fdabf19a82ff43449e052f

File: src/cpp/database/database.cpp [hunk 6]
```diff
@@ -1352,6 +1352,71 @@ std::vector<std::pair<EntityId, EntityId>> Database::get_entities_by_name(
...
+
+    // The mutex should be taken only once. get_entity_kind already locks.
+    std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+    for (auto reader : datareaders_[domain_id])
+    {
+        // Unlink related locators
+        for (auto locator : reader.second.get()->locators)
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/database/database.cpp]: Done in 20de40bddebda707251a80742ad5f816acef861e

File: src/cpp/database/database.cpp (hunk 1 above)
Comment (JLBuenoLopez-eProsima) [reply src/cpp/database/database.cpp]: The erase test has been completed to check that the maps are correctly erased in 460eaf2347095ad626848ba348254c32ef9e0cf9

The implementation has been changed in 13ff4e2fa03ae21d60a1fb12cb5dd20a00e0bf03

File: src/cpp/database/database.cpp (hunk 2 above)
Comment (JLBuenoLopez-eProsima) [reply src/cpp/database/database.cpp]: The regression test has been added in fd2487725036a37ab28dcb599c8bc4512e74e171

The implementation has been fixed in b525e9190608c5254df916184404b2c7942a6959

**Comment by JLBuenoLopez-eProsima** [comment]
Rebasing and refactoring `load_file` to pass Windows CI is waiting for #85 to be merged into `main`.

File: src/cpp/database/database.hpp [hunk 7]
```diff
@@ -135,8 +135,14 @@ class Database
      * @brief Erase all the data related to a domain.
      *
      * After the operation, the domain_id becomes invalid.
+     * Preconditions for this method are that:
+     *            * the domain must exist.
+     *            * the monitor attached to the domain must be stopped.
```
Comment (IkerLuengo) [inline src/cpp/database/database.hpp:140]:
```suggestion
     * @pre
     *            * the domain must exist.
     *            * the monitor attached to the domain must be stopped.
```

File: src/cpp/database/entities.hpp [hunk 8]
```diff
@@ -366,6 +359,8 @@ struct DataReader : DDSEndpoint
 
     /**
      * Clear the maps and data
+     * This method does not clear the locators map. Call DDSEndpoint::clear before to ensure that all the information
+     * is correctly disposed.
```
Comment (IkerLuengo) [inline src/cpp/database/entities.hpp:363]: `DDSEndpoint::clear` is no more

File: src/cpp/database/entities.hpp [hunk 9]
```diff
@@ -391,6 +386,8 @@ struct DataWriter : DDSEndpoint
...
```
Comment (IkerLuengo) [inline src/cpp/database/entities.hpp:390]: `DDSEndpoint::clear` is no more

File: test/unittest/Database/DatabaseEraseTests.cpp [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+    // Dump erased database
+    DatabaseDump erased_dump = db.dump_database();
+
+    // Compare erased and final dumps
+    EXPECT_EQ(final_dump, erased_dump);
+}
+
+// This test checks that erasing a database where the participant is not yet linked to the process works as expected
+TEST(database_erase_tests, erase_domain_unlinked_participant_process)
```
Comment (IkerLuengo) [inline test/unittest/Database/DatabaseEraseTests.cpp:120]: This is the same as `erase_domain` but with different dumps. We can extract this to a function that takes the dump names and make the tests call this function instead

**Review by IkerLuengo — CHANGES_REQUESTED** [review]

File: src/cpp/database/database.hpp (hunk 7 above)
Comment (JLBuenoLopez-eProsima) [reply src/cpp/database/database.hpp:140]: Good to know :wink:

**Review by IkerLuengo — COMMENTED** [review]
LGTM

**Review by IkerLuengo — APPROVED** [review]
:+1:

**Pull request merged**